File: MediaCrawler-Web/src/api/modules/dataDb.ts
import { httpDelete, httpGet } from '../request';
import type { DbPlatformMeta, DbQueryResult, TaskDataStats } from '@/types/config';

export function fetchDbPlatforms() {
  return httpGet<{ platforms: DbPlatformMeta[] }>('/api/data/db/platforms');
}

export function fetchDbData(
  platform: string,
  kind: string,
  params: { page?: number; page_size?: number; keyword?: string; content_id?: string; order_by?: string; order_direction?: string },
) {
  const q = new URLSearchParams();
  if (params.page != null) q.set('page', String(params.page));
  if (params.page_size != null) q.set('page_size', String(params.page_size));
  if (params.keyword) q.set('keyword', params.keyword);
  if (params.content_id) q.set('content_id', params.content_id);
  if (params.order_by) q.set('order_by', params.order_by);
  if (params.order_direction) q.set('order_direction', params.order_direction);
  const qs = q.toString();
  return httpGet<DbQueryResult>(`/api/data/db/${platform}/${kind}${qs ? `?${qs}` : ''}`);
}

export function fetchTaskData(
  platform: string,
  taskId: number,
  params: { page?: number; page_size?: number; order_by?: string; order_direction?: string },
) {
  const q = new URLSearchParams();
  if (params.page != null) q.set('page', String(params.page));
  if (params.page_size != null) q.set('page_size', String(params.page_size));
  if (params.order_by) q.set('order_by', params.order_by);
  if (params.order_direction) q.set('order_direction', params.order_direction);
  const qs = q.toString();
  return httpGet<DbQueryResult>(`/api/data/db/${platform}/task/${taskId}${qs ? `?${qs}` : ''}`);
}

export function fetchContentComments(
  platform: string,
  contentId: string,
  params: { page?: number; page_size?: number; order_by?: string; order_direction?: string },
) {
  const q = new URLSearchParams();
  if (params.page != null) q.set('page', String(params.page));
  if (params.page_size != null) q.set('page_size', String(params.page_size));
  if (params.order_by) q.set('order_by', params.order_by);
  if (params.order_direction) q.set('order_direction', params.order_direction);
  const qs = q.toString();
  return httpGet<DbQueryResult>(`/api/data/db/${platform}/comments/content/${contentId}${qs ? `?${qs}` : ''}`);
}

export function deleteDataRecord(platform: string, kind: string, recordId: number) {
  return httpDelete<{ status: string; message: string }>(`/api/data/db/${platform}/${kind}/${recordId}`);
}

export function fetchTaskDataStats(taskId: number) {
  return httpGet<TaskDataStats>(`/api/data/db/task/${taskId}/stats`);
}

export interface TaskInfo {
  task_id: number;
  keywords: string;
  status: string;
  created_at: string;
  record_count: number;
}

export function fetchAvailableTasks(platform: string, kind: string) {
  return httpGet<TaskInfo[]>(`/api/data/db/${platform}/${kind}/tasks`);
}

// ── 内容邻居（上一条/下一条）────────────────────────────────────────

export interface ContentNeighbors {
  platform: string;
  kind: string;
  current_content_id: string;
  prev: Record<string, unknown> | null;
  next: Record<string, unknown> | null;
}

export function fetchContentNeighbors(
  platform: string,
  params: {
    content_id: string;
    order_by?: string;
    order_direction?: string;
    keyword?: string;
    task_id?: number;
  },
) {
  const q = new URLSearchParams();
  q.set('content_id', params.content_id);
  if (params.order_by) q.set('order_by', params.order_by);
  if (params.order_direction) q.set('order_direction', params.order_direction);
  if (params.keyword) q.set('keyword', params.keyword);
  if (params.task_id != null) q.set('task_id', String(params.task_id));
  return httpGet<ContentNeighbors>(`/api/data/db/${platform}/contents/neighbors?${q.toString()}`);
}

File: MediaCrawler-Api/api/routers/data_db.py
# -*- coding: utf-8 -*-
from typing import Optional

from fastapi import APIRouter, HTTPException, Query

from services.data_query_service import DataQueryService, list_platforms

router = APIRouter(prefix="/data/db", tags=["data-db"])


@router.get("/platforms")
async def get_db_platforms():
    return {"platforms": list_platforms()}


@router.get("/task/{task_id}/stats")
async def get_task_data_stats(task_id: int):
    """获取任务的数据统计（各平台内容/评论数量）"""
    try:
        return await DataQueryService.get_task_data_stats(task_id)
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.get("/{platform}/task/{task_id}")
async def query_data_by_task(
    platform: str,
    task_id: int,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    """按任务 ID 查询该平台爬取的内容数据"""
    try:
        return await DataQueryService.query_by_task(platform, task_id, page, page_size)
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.get("/{platform}/{kind}/content/{content_id}")
async def query_comments_by_content(
    platform: str,
    kind: str,
    content_id: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    """按内容 ID 查询该内容下的评论"""
    if kind != "comments":
        raise HTTPException(400, "kind 必须为 'comments'，此端点仅用于查询评论")
    try:
        return await DataQueryService.query_comments_by_content(platform, content_id, page, page_size)
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.delete("/{platform}/{kind}/{record_id}")
async def delete_data_record(
    platform: str,
    kind: str,
    record_id: int,
):
    """删除单条数据记录"""
    try:
        deleted = await DataQueryService.delete_record(platform, kind, record_id)
        if not deleted:
            raise HTTPException(404, "记录不存在")
        return {"status": "ok", "message": "记录已删除"}
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.get("/{platform}/{kind}")
async def query_data(
    platform: str,
    kind: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    keyword: Optional[str] = None,
    content_id: Optional[str] = None,
):
    try:
        return await DataQueryService.query(platform, kind, page, page_size, keyword, content_id)
    except ValueError as e:
        raise HTTPException(400, str(e))
